### PR TITLE
RecurrentGemma: add generative tests

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Adds generative tests to `RecurrentGemma`, through the help of the `_is_stateful` flag (see #31478). A few changes were needed to handle failing tests, but we could remove many overwrites 🤗 

⚠️ `position_ids` was not being passed into the model at generation time, padding was broken

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Test changes look nice!

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -689,12 +689,14 @@ def _update_model_kwargs_for_generation(
                     dim=-1,
                 )
 
-        if (
-            model_kwargs.get("use_cache", True)
-            and "cache_position" in model_kwargs
-            and model_kwargs["cache_position"] is not None
-        ):
+        if model_kwargs.get("use_cache", True):
```
Comment (gante) [inline src/transformers/generation/utils.py:692]: Changes in this file copied from the [PR](https://github.com/huggingface/transformers/pull/31898) that updates `prepare_inputs_for_generation` (they are also needed for this PR)

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 2]
```diff
@@ -855,6 +855,7 @@ def forward(
         output_hidden_states = True
         outputs = self.model(
             input_ids=input_ids,
+            position_ids=position_ids,
```
Comment (gante) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:858]: this was missing -> generation with padding was broken 😬

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 3]
```diff
@@ -895,37 +896,31 @@ def forward(
...
     def prepare_inputs_for_generation(
-        self, input_ids, attention_mask=None, inputs_embeds=None, cache_position=None, use_cache=None, **kwargs
+        self, input_ids, attention_mask=None, cache_position=None, use_cache=None, position_ids=None, **kwargs
     ):
-        position_ids = kwargs.get("position_ids", None)
+        if cache_position.shape[0] == 1 and cache_position[0] == 0:
+            raise ValueError(
+                "Recurrent gemma doesn't work with an empty text prompt (i.e. only BOS token), as a minimum of two "
+                "tokens are needed to detect prefill. Please provide a non-empty text prompt."
+            )
```
Comment (gante) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:907]: We have to pick one, because we need to detect we're doing prefill in the fwd pass:
1. the model doesn't work with empty text prompt (we detect prefill with `cache_position.shape[0] != 1`)
2. the model doesn't compile (we detect prefill with `cache_position[0] == 0`)
3. we add a new flag in the forward pass (in the model class, before transforming `attention_mask`, we can detect prefill)

None of them seems to me like a clear win, so I went with the option that was already in the code :) This exception exposes the limitation to the user.


1. 
2.
3.

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 4]
```diff
@@ -895,37 +896,31 @@ def forward(
             hidden_states=outputs.hidden_states,
         )
 
-    # Ignore copy
+    # Ignore copy (no past_key_values, no inputs_embeds)
```
Comment (gante) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:899]: `inputs_embeds` was hitting corner cases, removed the support for now

File: tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py [hunk 5]
```diff
@@ -422,7 +387,20 @@ class RecurrentGemmaIntegrationTest(unittest.TestCase):
 
     @require_read_token
     def test_2b_generate(self):
-        EXPECTED_TEXTS = ['Hello I am doing a project on the topic of "The impact of the internet on the society" and I am looking for some information on the topic. I am looking for some information on the impact of the internet on the society. I am looking for some information on the impact of the internet on the society. I am looking for some', 'Hi today is a very good day for you. You will be able to do all the work you want to do. You will be able to do all the work you want to do. You will be able to do all the work you want to do. You will be able to do all the work you want to do.']  # fmt: skip
+        EXPECTED_TEXTS = [
```
Comment (gante) [inline tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py:390]: If we DON'T pass `position_ids` from `RecurrentGemmaForCausalLM` to the model itself, as before this PR, the tests pass without changes.

This update reflects the bugfix. Note that the 1st sequence (unpadded) doesn't change :)

**Comment by gante** [comment]
@ArthurZucker pushed further corrections. As per your last set of comments, the current diff doesn't break `torch.compile` on recurrent gemma  :)

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py (hunk 4 above)
Comment (ArthurZucker) [reply src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:899]: mmm this is breaking can't we keep input_embeds?

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:907]: interesting. Won't this be an issue for generate compile as this function would be compiled -> input dependent branches?

**Review by ArthurZucker — COMMENTED** [review]
Makes sense and good catches. We need to rebase !
